### DISCUSSION TITLE
Frequency domain analysis - two materials

### DISCUSSION THREAD

**Discussion by Luca-R27** [idaholab/moose#27071] [Q&A Modules: Solid mechanics] [answered]
Hello MOOSE users,

I am trying to modify the input script that performs a frequency-domain analysis of a cantilever to have it accept several materials. The source example is here:
modules/solid_mechanics/examples/wave_propagation/cantilever_sweep.i

I am reading a mesh from an exodus file where blocks were already pre-defined (blocks 1, 2, ... N). I would like to assign material properties based on the block definitions. The elastic properties is easy and having N snippets like this one works:
```
  [elastic_tensor_Al]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 68e9
    poissons_ratio = 0.36
    block = 1
  []
```
But I still haven't found a way to define densities based on block definitions. Density is provided in:
```
[Functions]
  [./freq2]
    type = ParsedFunction
    symbol_names = density
    symbol_values = 2.7e3 #Al kg/m3
    expression = '-t * t * density'
  [../]
[]
```
ParsedFunction does not accept a 'block' argument. I tried using variations of if-else statements, like " expression = if(block==1,'freq1' ,if(block==2, 'freq2', 'freq3')) ". I tried for expression and symbol_names/symbol_values but unsuccessfully. 

Any ideas?

Many thanks.
Luca

**Comment by GiudGiud** [comment]
Hello
Functions are global and predefined functions of space and time. They cannot be block restricted. 

What object type of the freq123?
Same question for t and density ?

If they are not functions or postprocessors, they cannot be used here. You ll have to use a ParsedMaterial. (Which can be block restricted)

guillaume

**Reply by Luca-R27** [reply]
Hi Guillaume,

Thanks for your reply. So copying from the original example in https://mooseframework.inl.gov/modules/solid_mechanics/1d_elastic_waves.html :
```
[Functions]
  [./freq2]
    type = ParsedFunction
    symbol_names = density
    symbol_values = 2.7e3 #Al kg/m3
    expression = '-t * t * density'
  [../]
[]

[Controls]
  [./func_control]
    type = RealFunctionControl
    parameter = 'Kernels/*/rate'
    function = 'freq2'
    execute_on = 'initial timestep_begin'
  [../]
[]
```
From here I understand that freq2 is of type FunctionName and density is the name (string) of a variable to which value 2700 is assigned inside the definition of freq2. t is time. This input file was among the examples provided in the solid_mechanics/examples folder. All I need to do is to specify -- based on blocks defined in the ingested mesh -- two different densities. So, if you like, it would be finding the resonances of a bilayered beam of say aluminium (say block = 1) and steel (say block = 2).
Since RealFunctionControl takes a function as parameter I cannot specify a ParsedMaterial instead of ParsedFunction, unless there's a workaround (?)


Thanks!
Luca

**Reply by GiudGiud** [reply, marked as answer]
Ok this is fine if t = time and density is a constant. 
The function will only be sampled at point 0,0,0 in a Control, so there's no point block restricting it. 

Maybe you should have multiple kernels, one for each block, and use one Control for each kernel

**Reply by Luca-R27** [reply]
This suggestion of multiple kernels and controls worked like a charm. Thanks!
